flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java

```diff
@@ -0,0 +1,790 @@
+
+    private final TaskManagerTracker taskManagerTracker;
+    private final ResourceTracker resourceTracker;
+    private final ResourceAllocationStrategy resourceAllocationStrategy;
+
+    private final SlotStatusSyncer slotStatusSyncer;
+
+    /** Scheduled executor for timeouts. */
+    private final ScheduledExecutor scheduledExecutor;
+
+    /** Timeout after which an unused TaskManager is released. */
+    private final Time taskManagerTimeout;
+
+    private final SlotManagerMetricGroup slotManagerMetricGroup;
+
+    private final Map<JobID, String> jobMasterTargetAddresses = new HashMap<>();
+
+    /** Defines the max limitation of the total number of task executors. */
+    private final int maxTaskManagerNum;
+
+    /** Defines the number of redundant task executors. */
+    private final int redundantTaskManagerNum;
+
+    /**
+     * Release task executor only when each produced result partition is either consumed or failed.
+     */
+    private final boolean waitResultConsumedBeforeRelease;
+
+    /** The default resource spec of workers to request. */
+    private final WorkerResourceSpec defaultWorkerResourceSpec;
+
+    /** The resource profile of default slot. */
+    private final ResourceProfile defaultSlotResourceProfile;
+
+    private boolean sendNotEnoughResourceNotifications = true;
+
+    /** ResourceManager's id. */
+    @Nullable private ResourceManagerId resourceManagerId;
+
+    /** Executor for future callbacks which have to be "synchronized". */
+    @Nullable private Executor mainThreadExecutor;
+
+    /** Callbacks for resource (de-)allocations. */
+    @Nullable private ResourceActions resourceActions;
+
+    private ScheduledFuture<?> taskManagerTimeoutsAndRedundancyCheck;
+
+    /** True iff the component has been started. */
+    private boolean started;
+
+    public FineGrainedSlotManager(
+            ScheduledExecutor scheduledExecutor,
+            SlotManagerConfiguration slotManagerConfiguration,
+            SlotManagerMetricGroup slotManagerMetricGroup,
+            ResourceTracker resourceTracker,
+            TaskManagerTracker taskManagerTracker) {
+
+        this.scheduledExecutor = Preconditions.checkNotNull(scheduledExecutor);
+
+        Preconditions.checkNotNull(slotManagerConfiguration);
+        this.taskManagerTimeout = slotManagerConfiguration.getTaskManagerTimeout();
+        this.redundantTaskManagerNum = slotManagerConfiguration.getRedundantTaskManagerNum();
+        this.waitResultConsumedBeforeRelease =
+                slotManagerConfiguration.isWaitResultConsumedBeforeRelease();
+        this.defaultWorkerResourceSpec = slotManagerConfiguration.getDefaultWorkerResourceSpec();
+        this.resourceAllocationStrategy = slotManagerConfiguration.getResourceAllocationStrategy();
+        int numSlotsPerWorker = slotManagerConfiguration.getNumSlotsPerWorker();
+        this.maxTaskManagerNum = slotManagerConfiguration.getMaxSlotNum() / numSlotsPerWorker;
```

Comment:
Not sure about converting max slot number to max TM number. This implicitly assume all TMs have the same number of slots.
I think we should convert max slot number to max total resources <cpu, memory>, and make sure that is not exceeded.
Given that this PR is already quite big. I would suggest to scope out the supporting for max slot number, make it a follow-up issue.

Reply:
I create https://issues.apache.org/jira/browse/FLINK-21177 for it.